v1.4.0 - SSR-2375 - Update User's Guide (#1099)
kintone-labs/kintone-ui-component: docs/document/website/versioned_docs/version-1.4.0/guides/version-conflicts-issue-solution.md

---
## Overview

> Version conflicts have been a problem for Kintone UI Component (KUC) users before the release of v1.4.0. This guide will help users understand the version conflicts issue and how to adapt and upgrade the Kintone UI Component package to the latest version starting from v1.4.0.

Reviewer: ```suggestion
Version conflicts have been a problem for Kintone UI Component (KUC) users before the release of v1.4.0. This guide will help users understand the version conflicts issue and how to adapt and upgrade the KUC package to the latest version starting from v1.4.0.
```

> We recommend that you should upgrade Kintone UI Component to the latest version for your projects.<br>

Reviewer: ```suggestion
We recommend that you should upgrade KUC to the latest version for your projects.<br>
```

```

### Using CDN

> ```sh
> https://unpkg.com/kintone-ui-component/umd/kuc.min.js
> ```
> As explained in the [Using UMD](#using-umd) section, you may still use `Kuc` as a global variable but be aware of the possible version conflicts errors that may occur.

Reviewer: How about this to make it simple?

```suggestion
See the guidance and sample code at [Use UMD](#using-umd) section. It's the same with using the `kuc.min.js` file.
```

### Using CDN

> Users using Kintone UI Component through the CDN using the below link will always have the latest version.<br>

Reviewer: ```suggestion
Users using KUC through the CDN using the below link will always have the latest version.<br>
```

Therefore, after the release of v1.4.0, some of your apps and plug-ins might not work correctly. <br>
```sh
https://unpkg.com/kintone-ui-component/umd/kuc.min.js
```
If you want to use a particular `Kuc` version, please specify the version number after the project name.
```sh
https://unpkg.com/kintone-ui-component@1.0.0/umd/kuc.min.js
```
As explained in the [Using UMD](#using-umd) section, you may still use `Kuc` as a global variable but be aware of the possible version conflicts errors that may occur.

### Using npm

> Users using the Kintone UI Component package through npm do not need to take any action but note the changes in how the components' tags and class names render. See the [Solution](#solution) section.

Reviewer: ```suggestion
Users using the KUC package through npm do not need to take any action but note the changes in how the components' tags and class names render. See the [Solution](#solution) section.
```

### Using CDN

Users using Kintone UI Component through the CDN using the below link will always have the latest version.<br>
Therefore, after the release of v1.4.0, some of your apps and plug-ins might not work correctly. <br>
```sh
https://unpkg.com/kintone-ui-component/umd/kuc.min.js
```
If you want to use a particular `Kuc` version, please specify the version number after the project name.
```sh
https://unpkg.com/kintone-ui-component@1.0.0/umd/kuc.min.js

> ```

Reviewer: I think you have a misunderstanding on CDN case.
unpkg automatically hosts kuc.min.js file after we release on npm. So the behavior and coding way is the same as UMD case above.
In case v1.4.0 is the latest version...
- This will be v1.4.0 kuc.min.js
  https://unpkg.com/kintone-ui-component/umd/kuc.min.js
- This will be v1.0.0 kuc.min.js
  https://unpkg.com/kintone-ui-component@1.0.0/umd/kuc.min.js

Thus, I think just linking to UMD is ok for CDN.

Sorry I didn't notice in the previous review.

### Using npm

Users using the Kintone UI Component package through npm do not need to take any action but note the changes in how the components' tags and class names render. See the [Solution](#solution) section.

## Caution required

### For users using versions before v1.4.0

Below are the common version conflicts errors that might occur:
#### 1. When importing multiple KUC packages and files (both ESM and UMD) of the same version or different versions:

In the example below, we load a v1.2.0 `kuc.min.js` file in the Kintone system and a v1.3.0 `kuc.min.js` file in our app. An `Illegal constructor` error will show when we try to create a new KUC Button component.

![Illegal constructor error when importing multiple kuc.min.js files](assets/UMD_multi_files.jpeg)

#### 2. When importing multiple KUC packages (ESM) and using custom HTML tags directly:

For example, we have two scripts as follow:
- One script uses v1.2.0 and specifies the text color of Dropdown to be green.
- One script uses v1.3.1 and specifies the text color of Dropdown to be red

> In the first case, we have the v1.2.0 script loaded first and v1.3.1 loaded last. We can see that the text color of Dropdown in the v1.3.1 script is overridden by v1.2.0 script.

Reviewer: How about unifying the version to use with 1. case? (v1.2.0 and v1.3.0)
About the image, you can override only the text area!

![Custom element is overridden by the first loaded file](assets/ESM_multi_files_1.png)

Now, if the order of import is switched, and the v1.3.1 is loaded first. We can see the text color of Dropdown in the v1.2.0 script is overridden by v1.3.1 script.

![Custom element is overridden by the first loaded file](assets/ESM_multi_files_2.png)

### For users using version v1.4.0 and above

After the release of v1.4.0, loading multiple KUC packages (ESM) of the same version or different versions will not cause version conflicts errors anymore.

> However, please note that when loading multiple `kuc.min.js` files (UMD), only **the last loaded** one is used. Therefore, if you use `Kuc` instead of the recommended `Kucs` object, please note that it will refer to the last loaded `kuc.min.js` file and that may not be the version you wanted.

Reviewer: Maybe there is a misunderstanding.
I think the below is correct. How do you think?

```suggestion
However, please note that when loading multiple `kuc.min.js` files (UMD), only **the last loaded** one is used if you use `Kuc` instead of the recommended `Kucs` object. Please note that it may not be the version you wanted.
```

> The following example import orders will not cause version conflicts errors:

Reviewer: How about adding this sentence?

```suggestion
When there are both versions before and on/after v1.4.0, you need to be careful.<br>

The following example import orders will not cause version conflicts errors:
```
